### PR TITLE
[performance](stream-load) use vector instead of skiplist when insert dup keys

### PR DESCRIPTION
**Pull request by hust-hhb**
# Proposed changes

Issue Number: close #xxx

## Problem summary

when insert dup keys, use vector instead of skiplist will reduce 50% loading time.
Here is my test on tpch lineitem table with 3 replica, before change code, loading 6,000,000 rows cost 12s.
{
    "TxnId": 126,
    "Label": "b00c488f-4bd4-4673-bbff-a6f32f5d2937",
    "Comment": "",
    "TwoPhaseCommit": "false",
    "Status": "Success",
    "Message": "OK",
    "NumberTotalRows": 5999444,
    "NumberLoadedRows": 5999444,
    "NumberFilteredRows": 0,
    "NumberUnselectedRows": 0,
    "LoadBytes": 778778561,
    "LoadTimeMs": 12034,
    "BeginTxnTimeMs": 0,
    "StreamLoadPutTimeMs": 1,
    "ReadDataTimeMs": 9135,
    "WriteDataTimeMs": 12006,
    "CommitAndPublishTimeMs": 26
}
After using vector and sort it on flush stage, it only cost 6.4s.
{
    "TxnId": 481,
    "Label": "21b3634d-ab7c-4813-885e-7f6215111e7c",
    "Comment": "",
    "TwoPhaseCommit": "false",
    "Status": "Success",
    "Message": "OK",
    "NumberTotalRows": 6001215,
    "NumberLoadedRows": 6001215,
    "NumberFilteredRows": 0,
    "NumberUnselectedRows": 0,
    "LoadBytes": 771862139,
    "LoadTimeMs": 6434,
    "BeginTxnTimeMs": 0,
    "StreamLoadPutTimeMs": 1,
    "ReadDataTimeMs": 4718,
    "WriteDataTimeMs": 6409,
    "CommitAndPublishTimeMs": 23
}

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhannngchen** [comment]
run buildall

File: be/src/olap/memtable.cpp [hunk 1]
```diff
@@ -243,11 +243,16 @@ void MemTable::_collect_vskiplist_results() {
     VecTable::Iterator it(_vec_skip_list.get());
     vectorized::Block in_block = _input_mutable_block.to_block();
     if (_keys_type == KeysType::DUP_KEYS) {
+        RowComparator _cmp(_schema);
+        vectorized::MutableBlock mutable_block =
+                vectorized::MutableBlock::build_mutable_block(&in_block);
+        _cmp.set_block(&mutable_block);
+        std::sort(_vec_row.begin(), _vec_row.end(), _cmp);
```
Comment (zhannngchen) [inline be/src/olap/memtable.cpp:250]: you can use the `_vec_row_comparator` with lambda directly, like this way:
```
        std::sort(_vec_row.begin(), _vec_row.end(),
                  [this](const RowInBlock* l, const RowInBlock* r) -> bool {
                      return (*(this->_vec_row_comparator))(l, r) < 0;
                  });
```

File: be/src/olap/memtable.cpp [hunk 2]
```diff
@@ -191,8 +191,8 @@ void MemTable::_insert_one_row_from_block(RowInBlock* row_in_block) {
     _rows++;
     bool overwritten = false;
     if (_keys_type == KeysType::DUP_KEYS) {
-        // TODO: dup keys only need sort opertaion. Rethink skiplist is the beat way to sort columns?
-        _vec_skip_list->Insert(row_in_block, &overwritten);
+        // dup keys only need sort operation, use vector to sort is faster than skiplist
+        _vec_row.emplace_back(row_in_block);
```
Comment (zhannngchen) [inline be/src/olap/memtable.cpp:195]: reserve space at `insert` will be faster.

File: be/src/olap/memtable.h [hunk 3]
```diff
@@ -88,6 +88,26 @@ class MemTable {
...
+                                       *_pblock, -1);
+        }
+
+        bool operator()(const RowInBlock* left, const RowInBlock* right) const {
+            return compare(left, right) < 0;
+        }
+
+    private:
+        const Schema* _schema;
+        vectorized::MutableBlock* _pblock; // 对应Memtable::_input_mutable_block
```
Comment (carlvinhust2012) [inline be/src/olap/memtable.h:108]: here may be use English comment is more better

**Comment by hust-hhb** [comment]
@zhannngchen @carlvinhust2012 Thanks for comment, now i use _row_in_blocks to store and sort dup keys, which is already exist in memtable, it no need to create a new vector and RowComparator to do this.

**Comment by hust-hhb** [comment]
run buildall

File: be/src/olap/memtable.cpp [hunk 4]
```diff
@@ -182,17 +182,16 @@ void MemTable::insert(const vectorized::Block* input_block, const std::vector<in
     _mem_usage += input_size;
     _insert_mem_tracker->consume(input_size);
     for (int i = 0; i < num_rows; i++) {
-        _row_in_blocks.emplace_back(new RowInBlock {cursor_in_mutableblock + i});
-        _insert_one_row_from_block(_row_in_blocks.back());
+        _row_in_blocks.emplace(_row_in_blocks.begin(), new RowInBlock {cursor_in_mutableblock + i});
```
Comment (zhannngchen) [inline be/src/olap/memtable.cpp:185]: Why insert elements to the head of `_row_in_blocks`? It's quite low efficient for vector

**Comment by hust-hhb** [comment]
run buildall

File: be/src/olap/memtable.cpp [hunk 5]
```diff
@@ -191,8 +191,7 @@ void MemTable::_insert_one_row_from_block(RowInBlock* row_in_block) {
     _rows++;
     bool overwritten = false;
     if (_keys_type == KeysType::DUP_KEYS) {
-        // TODO: dup keys only need sort opertaion. Rethink skiplist is the beat way to sort columns?
-        _vec_skip_list->Insert(row_in_block, &overwritten);
+        // dup keys store in vector _row_in_blocks and sort it on flush stage
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:194]: store row_in_blocks in vector.

File: be/src/olap/memtable.cpp [hunk 6]
```diff
@@ -243,11 +242,22 @@ void MemTable::_collect_vskiplist_results() {
...
     if (_keys_type == KeysType::DUP_KEYS) {
+        vectorized::MutableBlock mutable_block =
+                vectorized::MutableBlock::build_mutable_block(&in_block);
+        _vec_row_comparator->set_block(&mutable_block);
+        std::sort(_row_in_blocks.begin(), _row_in_blocks.end(),
+                  [this](const RowInBlock* l, const RowInBlock* r) -> bool {
+                      if ((*(this->_vec_row_comparator))(l, r) == 0) {
+                          return l->_row_pos - r->_row_pos > 0;
+                      } else {
+                          return (*(this->_vec_row_comparator))(l, r) < 0;
```
Comment (dataroaring) [inline be/src/olap/memtable.cpp:253]: two times comparator?

**Review by dataroaring — CHANGES_REQUESTED** [review]
add order by for case which does not generate a stable out.

File: be/src/olap/memtable.cpp (hunk 6 above)
Comment (zhannngchen) [reply be/src/olap/memtable.cpp:253]: +1

**Comment by hust-hhb** [comment]
run buildall

File: be/src/olap/memtable.cpp [hunk 7]
```diff
@@ -243,11 +242,23 @@ void MemTable::_collect_vskiplist_results() {
     VecTable::Iterator it(_vec_skip_list.get());
     vectorized::Block in_block = _input_mutable_block.to_block();
     if (_keys_type == KeysType::DUP_KEYS) {
+        vectorized::MutableBlock mutable_block =
+                vectorized::MutableBlock::build_mutable_block(&in_block);
+        _vec_row_comparator->set_block(&mutable_block);
+        std::sort(_row_in_blocks.begin(), _row_in_blocks.end(),
```
Comment (liaoxin01) [inline be/src/olap/memtable.cpp:248]: Can we use `std::stable_sort` here directly? 
 `std::stable_sort(_row_in_blocks.begin(), _row_in_blocks.end(), _vec_row_comparator);`

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**